File: arvBinaria.cpp
#include <iostream>
#include <Windows.h>
using namespace std;

class No
{
public:
    No *left, *right;
    int key;

    No(int key)
    {
        this->key = key;
        left = NULL;
        right = NULL;
    }

    int getKey()
    {
        return key;
    }

    No *getLeft()
    {
        return left;
    }

    No *getRight()
    {
        return right;
    }

    void setLeft(No *no)
    {
        left = no;
    }
    void setRight(No *no)
    {
        right = no;
    }
};

class Tree
{
private:
    No *root;

public:
    Tree()
    {
        root = NULL;
    }

    No *getRoot()
    {
        return root;
    }

    void insert(int key)
    {
        if (root == NULL)
            root = new No(key);
        else
        {
            insertAux(root, key);
        }
        cout<<"Nó " << key << " criado"<<endl;
    }

    void insertAux(No *no, int key)
    {
        if (key < no->getKey())
        {
            if (no->getLeft() == NULL)
            {
                No *new_no = new No(key);
                no->setLeft(new_no);
            }
            else
            {
                insertAux(no->getLeft(), key);
            }
        }
        else if (key > no->getKey())
        {
            if (no->getRight() == NULL)
            {
                No *new_no = new No(key);
                no->setRight(new_no);
            }
            else
            {
                insertAux(no->getRight(), key);
            }
        }
    }

    void inOrder(No *no)
    {
        if (no != NULL)
        {
            inOrder(no->getLeft());
            cout << no->getKey() << " | ";
            inOrder(no->getRight());
        }
    }

    int height(No *no)
    {

        if (no == NULL)
            return 0;
        else
        {
            /* compute the depth of each subtree */
            int left = height(no->getLeft());
            int right = height(no->getRight());

            /* use the larger one */
            if (left > right)
                return (left + 1);
            else
                return (right + 1);
        }
        cout << endl;
    }

    int quantNos(No *no)
    {
        if (no == NULL)
            return 0;
        else
            return 1 + quantNos(no->getLeft()) + quantNos(no->getRight());

        cout << endl;
    }

    int quantLeaves(No *no)
    {
        if (no == NULL)
            return (0);
        else if ((no->getLeft()) == NULL && (no->getRight()) == NULL)
            return (1);
        else
            return (quantLeaves(no->getLeft()) + quantLeaves(no->getRight()));
    }

    No* noRemove(No* r, int v)
    {
        if (r == NULL)
            return NULL;
        else if (r->getKey() > v)
            r->setLeft(noRemove(r->getLeft(), v));
        else if (r->getKey() < v)
            r->setRight(noRemove(r->getRight(), v));
        else
        {   /* achou o nó a remover */
            /* nó sem filhos */
            if (r->getLeft() == NULL && r->getRight() == NULL)
            {
                free(r);
                r = NULL;
                cout<<"Nó " << v << " deletado"<<endl;
            }
            /* nó só tem filho à direita */
            else if (r->getLeft() == NULL)
            {
                No* t = r;
                r = r->getRight();
                free(t);
                cout<<"Nó " << v << " deletado"<<endl;
            }
            /* só tem filho à esquerda */
            else if (r->getRight() == NULL)
            {
                No* t = r;
                r = r->getLeft();
                free(t);
                cout<<"Nó " << v << " deletado"<<endl;
            }
            /* nó tem os dois filhos */
            else
            {
                No* f = r->left;
                while (f->right != NULL)
                {
                    f = f->right;
                }
                r->key = f->key; /* troca as informações */
                f->key = v;
                r->left = noRemove(r->left, v);
                cout<<"Nó " << v << " deletado"<<endl;
            }
        }
        return r;
    }
};

int main()
{
SetConsoleOutputCP(65001);

    Tree tree;
    int o, x;

    while (o != 7)
    {
        cout << "===============================\n";
        cout << "1 - Inserir;\n";
        cout << "2 - Altura;\n";
        cout << "3 - Quantidade de nós;\n";
        cout << "4 - Quantidade de folhas;\n";
        cout << "5 - remover;\n";
        cout << "6 - Imprimir arvore;\n";
        cout << "7 - Sair.\n";
        cout << "===============================\n";
        cout << "Selecione a opção de escolha: ";
        cin >> o;
        switch (o)
        {
        case 1:
            /* cout << "Digite o valor: ";
            cin >> x;
            tree.insert(x);
            break; */

            tree.insert(8);
            tree.insert(4);
            tree.insert(2);
            tree.insert(6);
            tree.insert(1);
            tree.insert(3);
            tree.insert(5);
            tree.insert(12);
            tree.insert(10);
            tree.insert(14);
            tree.insert(11);
            tree.insert(13);
            tree.insert(15);
            break;

        case 2:
            cout << "Altura da arvore: " << tree.height(tree.getRoot()) << endl;
            break;

        case 3:
            cout << "Quantidade de nós: " << tree.quantNos(tree.getRoot()) << endl;
            break;

        case 4:
            cout << "Quantidade de folhas: " << tree.quantLeaves(tree.getRoot()) << endl;
            break;

        case 5:
            int rm;
            cout<<"Digite qual nó deseja remover: ";
            cin>> rm;
            tree.noRemove(tree.getRoot(), rm);
            break;

        case 6:
            cout << "Imprimindo arvore em ordem \n";
            tree.inOrder(tree.getRoot());
            cout <<endl;
            cout<<"Raiz: " << tree.getRoot()->getKey()<<endl;
            break;

        default:
            break;
        }
    }
}

File: arvBinaria.py
class Node:

    def __init__(self, label):
        self.label = label
        self.left = None
        self.right = None

    def getLabel(self):
        return self.label

    def setLabel(self, label):
        self.label = label

    def getLeft(self):
        return self.left

    def setLeft(self, left):
        self.left = left

    def getRight(self):
        return self.right

    def setRight(self, right):
        self.right = right


class BinarySearchTree:

    def __init__(self):
        self.root = None

    def insert(self, label):

        # cria um novo nó
        node = Node(label)

        # verifica se a árvore está vazia
        if self.empty():
            self.root = node
        else:

            # árvore não vazia, insere recursivamente

            dad_node = None
            curr_node = self.root

            while True:
                if curr_node != None:
                    dad_node = curr_node

                    # verifica se vai para esquerda ou direita
                    if node.getLabel() < curr_node.getLabel():
                        # vai para esquerda
                        curr_node = curr_node.getLeft()
                    else:
                        # vai para direita
                        curr_node = curr_node.getRight()
                else:

                    # se curr_node é None, então encontrou onde inserir
                    if node.getLabel() < dad_node.getLabel():
                        dad_node.setLeft(node)
                    else:
                        dad_node.setRight(node)

                    break  # sai do loop
        print("Nó " + str(label) + " criado!")

    def empty(self):
        if self.root == None:
            return True
        return False

    # mostra em pré-ordem (raiz-esquerda-direita)
    def show(self, curr_node):
        if curr_node != None:
            print('%d' % curr_node.getLabel(), end=' ')
            self.show(curr_node.getLeft())
            self.show(curr_node.getRight())

    def getRoot(self):
        return self.root

    def height(self, no):

        if no == None:
            return 0
        else:
            left = self.height(no.getLeft())
            right = self.height(no.getRight())

            if left > right:
                return (left + 1)
            else:
                return (right + 1)

    def quantNos(self, no):

        if no == None:
            return 0
        else:
            return 1 + self.quantNos(no.getLeft()) + self.quantNos(no.getRight())

    def quantLeaves(self, no):

        if no == None:
            return (0)
        elif ((no.getLeft()) == None and (no.getRight()) == None):
            return (1)
        else:
            return (self.quantLeaves(no.getLeft()) + self.quantLeaves(no.getRight()))

    def noRemove(self, r, v):
        if (r == None):
            return None
        elif (r.getLabel() > v):
            r.setLeft(self.noRemove(r.getLeft(), v))
        elif (r.getLabel() < v):
            r.setRight(self.noRemove(r.getRight(), v))
        else:
            #achou o nó a remover
            #nó sem filhos
            if (r.getLeft() == None and r.getRight() == None):
            
                r = None
                print("Nó " + str(v) + " deletado")
            
            #nó só tem filho à direita
            elif (r.getLeft() == None):
                t = r
                r = r.getRight()
                del t
                print("Nó " + str(v) + " deletado")
            
            #só tem filho à esquerda
            elif (r.getRight() == None):
            
                t = r
                r = r.getLeft()
                del t 
                print("Nó " + str(v) + " deletado")
            
            #nó tem os dois filhos
            else:
            
                f = r.getLeft()
                while (f.getRight() != None):
                
                    f = f.getRight()
                
                r.label = f.label #troca as informações
                f.label = v
                r.setLeft(self.noRemove(r.getLeft(), v))
                print("Nó " + str(v) + " deletado")
            
        
        return r
    

t = BinarySearchTree()
o = None

while o != 7:
    op = int(input("\n===============================\n1 - Inserir;\n2 - Altura;\n3 - Quantidade de nós;\n4 - Quantidade de folhas;\n5 - remover;\n6 - Imprimir arvore;\n7 - Sair.\n===============================\nSelecione a opção de escolha: "))

    if op == 1:
        #valor = int(input("Digite o valor do nó: "))
        # t.insert(valor)

        t.insert(8)
        t.insert(4)
        t.insert(2)
        t.insert(6)
        t.insert(1)
        t.insert(3)
        t.insert(5)
        t.insert(12)
        t.insert(10)
        t.insert(14)
        t.insert(11)
        t.insert(13)
        t.insert(15)

    elif op == 2:
        print("Altura da arvore: " + str(t.height(t.getRoot())))
    elif op == 3:
        print("Quantidade de nós: " + str(t.quantNos(t.getRoot())))
    elif op == 4:
        print("Quantidade de folhas: " + str(t.quantLeaves(t.getRoot())))
    elif op == 5:
        rm = int(input("Digite qual nó deseja remover: "))
        t.noRemove(t.getRoot(), rm)
    elif op == 6:
        t.show(t.getRoot())
    else:
        break
